tools/gen-magic-strings.py

```diff
@@ -0,0 +1,240 @@
+                guards = guard_stack[-1]
+                guards[-1] = '!(%s)' % guards[-1]
+                guards.append(process_guard(elif_match.group(1)))
+            elif else_match is not None:
+                guards = guard_stack[-1]
+                guards[-1] = '!(%s)' % guards[-1]
+            elif endif_match is not None:
+                guard_stack.pop()
+
+            process_line(fname, lnum + 1, line, guard_stack)
+
+        if len(guard_stack) > 0:
+            print('warning: {fname}: unbalanced preprocessor conditional '
+                  'directives (analysis finished with no closing `#endif` '
+                  'for {guard_stack})'
+                  .format(fname=fname, guard_stack=guard_stack))
+
+    for root, dirs, files in os.walk(os.path.join(PROJECT_DIR, 'jerry-core')):
+        for fname in files:
+            if (fname.endswith('.c') or fname.endswith('.h')) \
+               and fname != 'lit-magic-strings.inc.h':
+                process_file(os.path.join(root, fname))
+
+    return results
+
+
+def calculate_magic_string_definition_guards(defs, uses):
+    extended_defs = []
+
+    for str_ref, str_value in defs:
+        if str_ref not in uses:
+            print('warning: unused magic string {str_ref}'
+                  .format(str_ref=str_ref))
+            continue
+
+        # calculate the most compact guard
+        guards = [set(guard_tuple)
+                  for guard_tuple, file_lns in uses[str_ref].items()]
+        for i in range(len(guards)):
+            if guards[i] is None:
+                continue
+            for j in range(len(guards)):
+                if j == i or guards[j] is None:
+                    continue
+                if guards[i] < guards[j]:
+                    guards[j] = None
+        guards = set(tuple(sorted(guard))
+                     for guard in guards if guard is not None)
+
+        extended_defs.append((str_ref, str_value, guards))
+
+    return extended_defs
+
+
+def guards_to_str(guards):
+    return ' \\\n|| '.join(' && '.join(g for g in sorted(guard))
+                          for guard in sorted(guards))
+
+
+def generate_header(f):
+    header = \
+"""/* Copyright JS Foundation and other contributors, http://js.foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * List of ECMA magic strings
+ *
+ * Generated by tools/gen-magic-strings.py
+ * DO NOT EDIT!!!
+ */"""
+    print(header, file=f)
```

Comment:
It was not trivial, but now the script has the same behaviour on py2.7 and py3.6